### ISSUE TITLE
Inconsistent size between icons and fonts in statusbar

### ISSUE THREAD

**Issue by 766F6964** [bakkeby/dwm-flexipatch#255]
I am unsure if this is a problem that is specific to dwm-flexipatch, dwm in general or the fonts that I am using. 
So, if the issue is out of place, feel free to let me know.
Ever since I started using dwm, I had issues getting icons and text in the status bar to show up with the same size.

I am running dwm-flexipatch and I am using slstatus for the statusbar.
The font is specified in the dwm config.h file.
The config really isn't special, I just specified: `static const char *fonts[] = { "SauceCodePro Nerd Font Mono:size=10:style=bold" };` (I also tried with `style=regular` instead of bold, same problem)

I tried various fonts (JetBrainMono Nerd Fonto Mono, SauceCodePro Nerd Font Mono, Material-Icons Font ...) but with all of them, I experience the same problem. The icons and the text do not have the same size:

![2022-05-31_01-17](https://user-images.githubusercontent.com/34845270/171067253-c48aff35-2ed3-484d-a2c5-08bcf1a70cf3.png)

As illustrated in the picture, the letters are bigger than the icons. I also made sure the icons are NerdFont Icons that are contained in my font (I copied them from the NerdFonts website: https://www.nerdfonts.com/cheat-sheet)

I tried all kinds of fonts, all have the same issue. And unfortunately its not possible to adjust the size for icons and text independently, if I increase the font-size, the issue remains.

Is there a patch that I am missing? What causes this issue, and is there a way to solve this?

**Comment by bakkeby** [comment]
Not all fonts are created equal and this is a typical issue when mixing different fonts. In this case you also have fonts that are patched with the Nerd Fonts icons.

A tool that I like to use is the GNOME Character map (gucharmap) where you can see the icons for one specific font only by going to View > Show only glyphs from this font.

![image](https://user-images.githubusercontent.com/1438008/171109790-4de1d8fb-76b6-46e5-a1f2-d4256e23fcbf.png)

What you may observe is that the icons often have additional spacing around them compared to a large character like W for example.

In dwm `fonts` is an array of fonts where the first font defined is the primary font and the rest are fallback fonts. The size of the primary font also governs the size of additional fonts that are loaded in later that contains glyphs that are not covered by any of the fonts listed in the fonts array on startup.

To work around the issue of Nerd Font icons being smaller than you'd like what you can do is to first select the primary font that you want to use for regular characters, A-Z etc., then use the Nerd Font as a fallback. This will allow you to control the size of the two independently. Additionally what you may want to try is to use pixelsize rather than size.

Assuming Sauce Code Pro is your font of choice you could have something like this.

```c
static const char *fonts[] = {
        "SauceCodePro Mono:pixelsize=14:style=bold",
        "SauceCodePro Nerd Font Mono:pixelsize=16:style=bold",
 };
```

**Comment by 766F6964** [comment]
Thanks for the insight.
What I also found interesting that it seems that the _Mono_ variant of the NerdFonts causes these issues. When I switched to the Non-Mono variant only, I did not have these problems, and the size is consistent:
And all that by only defining one font in my config.

![2022-05-31_09-17](https://user-images.githubusercontent.com/34845270/171114967-95d3f32b-4973-47ec-861f-0f1e2bd578f4.png)

```
static const char *fonts[] = { "SauceCodePro Nerd Font:size=10:antialias=true:autohint=true:style=Bold" };
```

Its good to know that sizes can be configured independently when having two fonts defined - however, that requires that the first font does not already contain the icons - because then the second font would never be used.
And for NerdFonts that is usually the case, unless special patching was done.

**Comment by bakkeby** [comment]
> however, that requires that the first font does not already contain the icons

Yes that is a given, but as the font is patched it should normally not be a problem getting hold of the unpatched version of the font.

As for Mono variant vs non-Mono variant, I just checked and the icon size in the non-Mono variant is generally bigger than in the mono one.

Mono:
![image](https://user-images.githubusercontent.com/1438008/171118615-6c5aa0f0-4578-49d8-ab66-73a513710061.png)

non-Mono:
![image](https://user-images.githubusercontent.com/1438008/171118716-ddc65e38-0f8d-4fef-9bb4-8f60e0604780.png)

**Comment by bakkeby** [comment]
Adding to this it seems that when using a non-mono Nerd Font the underlying Xft library can't always get an accurate read on the width of the icon / symbol / glyph. As an example using non-mono Nerd Font symbols for the tag icons can result in the icon not being centered within the tag on the bar, whereas the same symbols using a mono Nerd Font will be centered.

It is not yet clear to me whether this has to do with how the icon set is patched into the font or if it has something to do with the Xft library.

**Comment by AlexDumanskiy** [comment]
In my case, it helped to use "pixelsize" instead of "size". Same font, but with "pixelsize" the icons appear of normal proportions, while being ugly small with "size".
I have no idea how they work, so you can try this if you get as desperate as me.
